nova_vm/src/ecmascript/builtins/numbers_and_dates/math_object.rs

```diff
@@ -756,51 +756,126 @@ impl MathObject {
             return Ok(Value::neg_infinity());
         }
         // 4. If n < -1𝔽, return NaN.
+        if n.is_sign_negative(agent) {
+            return Ok(Value::nan());
+        }
         // 5. Return an implementation-approximated Number value representing the natural logarithm of 1 + ℝ(n).
+        Ok(Value::from_f64(agent, n.into_f64(agent).ln_1p()))
     }
 
     fn log10(agent: &mut Agent, _this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
         // 1. Let n be ? ToNumber(x).
+        let n = to_number(agent, arguments.get(0))?;
         // 2. If n is either NaN or +∞𝔽, return n.
+        if n.is_nan(agent) || n.is_pos_infinity(agent) {
+            return Ok(n.into_value());
+        }
         // 3. If n is 1𝔽, return +0𝔽.
+        if n.is_pos_one(agent) {
+            return Ok(Value::zero());
+        }
         // 4. If n is either +0𝔽 or -0𝔽, return -∞𝔽.
+        if n.is_pos_zero(agent) || n.is_neg_zero(agent) {
+            return Ok(Value::neg_infinity());
+        }
         // 5. If n < -0𝔽, return NaN.
+        if n.is_sign_negative(agent) {
+            return Ok(Value::nan());
+        }
+        
         // 6. Return an implementation-approximated Number value representing the base 10 logarithm of ℝ(n).
+        return Ok(Value::from_f64(agent, n.into_f64(agent).log10()))
     }
 
     fn log2(agent: &mut Agent, _this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
         // 1. Let n be ? ToNumber(x).
+        let n = to_number(agent, arguments.get(0))?;
         // 2. If n is either NaN or +∞𝔽, return n.
+        if n.is_nan(agent) || n.is_pos_infinity(agent) {
+            return Ok(n.into_value());
+        }
         // 3. If n is 1𝔽, return +0𝔽.
+        if n.is_pos_one(agent) {
+            return Ok(Value::zero());
+        }
         // 4. If n is either +0𝔽 or -0𝔽, return -∞𝔽.
+        if n.is_pos_zero(agent) || n.is_neg_zero(agent) {
+            return Ok(Value::neg_infinity());
+        }
         // 5. If n < -0𝔽, return NaN.
+        if n.is_sign_negative(agent) {
+            return Ok(Value::nan());
+        }
         // 6. Return an implementation-approximated Number value
+        Ok(Value::from_f64(agent, n.into_f64(agent).log2()))
     }
 
     fn max(agent: &mut Agent, _this_value: Value, arguments: ArgumentsList) -> JsResult<Value> {
         // 1. Let coerced be a new empty List.
+        let mut coerced = vec![];
```

Comment:
**nitpick:** `Vec::with_capacity(arguments.len())`

Reply:
Or you can actually do `let coerced = arguments.iter().map(...).collect::<Vec<_>>();`